File: dnd.js
class Character {
    constructor(_name, _charClass, _weapon, _race) {
        this.name = _name
        this.charClass = _charClass
        this.weapon = _weapon
        this.race = _race

        this.health = 100
        this.specialSkill

        if(this.charClass === 'warlock') {
            this.specialSkill = 'fireball'
        } else if(this.charClass === 'rogue') {
            this.specialSkill = 'backstab'
        } else {
            this.specialSkill = 'head punch'
            
        }
        
    }
    
    greeting() {
        // console.log('Hello, my name is ' + this.name + 'I am a ' + this.race + ' ' + this.charClass)
        console.log(`Hello, my name is ${this.name} I am a ${this.race} ${this.charClass}`);
    }
    
}

class Tavern {
    constructor(_tavernName, _hasInn, _questBoard, _city) {
        this.tavernName = _tavernName;
        this.hasInn = _hasInn;
        this.questBoard = _questBoard;
        this.city = _city;
    }

    rest(character) {
        if(this.hasInn === true) {
            character.health += 50
        }
    }
}

let ryan = new Character('ryan', 'warlock', 'staff', 'muggle');
let tavern = new Tavern('The Tavern', true, false, 'mordor')

// console.log(ryan)

// ryan.greeting()

// console.log(`${ryan.name} uses ${ryan.specialSkill}`)

ryan.health = 25

console.log(ryan.health)

tavern.rest(ryan)

console.log(ryan.health)

// let person = {
//     name: 'tim',
//     age: 25,
//     hair_color: 'brown'
// }

// // console.log(person)

// class Car {
//     constructor(_make, _model, _color) {
//         this.make = _make
//         this.model = _model
//         this.color = _color
//     }

// }
// let factory = []

// let truck1 = new Car('chevy', 'silverado', 'white')
// console.log(truck1.model)

File: dnd.go
package main

import "fmt"

/////////////////////////////////////////////////////////////////////////////////////
// CHARACTER
/////////////////////////////////////////////////////////////////////////////////////

// FAUX CLASS DATA TYPE CHARACTER
type Character struct {
	name string
	charClass string
	weapon string
	race string
	
	health int
	skillSlot int
	level int
	
	// backpack []string
}

// FAUX CONSTRUCTOR FOR STRUCT DATA TYPE  //  FOR CHARACTER
func NewCharacter(_name, _charClass, _weapon, _race string) *Character {
	// return Character{_name, _charClass, _weapon, _race, 100, 2, 1, items()}
	return &Character{_name, _charClass, _weapon, _race, 100, 2, 1}
}

// FUNCTIONS OR METHODS THAT BELONG TO THE CHARACTER DATA TYPE
func (c Character) greeting() {
	fmt.Println("Hello", c.name)
}

func (c *Character) dmg() {
	fmt.Println(c.name, "takes 50 damage")
	c.health -= 50
}

/////////////////////////////////////////////////////////////////////////////////////
// CHARACTER END
/////////////////////////////////////////////////////////////////////////////////////


/////////////////////////////////////////////////////////////////////////////////////
// TAVERN
/////////////////////////////////////////////////////////////////////////////////////

// FAUX CLASS DATA TYPE TAVERN
type Tavern struct {
	tavernName string
	hasInn bool
	questBoard bool
	city string
}

// FAUX CONSTRUCTOR FOR STRUCT DATA TYPE  //  FOR TAVERN
func NewTavern(tavernName string, hasInn bool, questBoard bool, city string) Tavern {
	return Tavern{tavernName, hasInn, questBoard, city}
}

// FUNCTIONS OR METHODS THAT BELONG TO THE CHARACTER DATA TYPE
func (t Tavern) rest(character *Character) {
	if t.hasInn == true {
		character.health += 25
	}
}

/////////////////////////////////////////////////////////////////////////////////////
// TAVERN END
/////////////////////////////////////////////////////////////////////////////////////

func main() {
	ryanDavis := NewCharacter("ryan", "warlock", "staff", "muggle")
	tavern := NewTavern("The Dodgey Tavern", true, true, "mordor")
	ryanDavis.greeting()
	fmt.Println(ryanDavis.health)
	ryanDavis.dmg()
	fmt.Println(ryanDavis.health)
	tavern.rest(ryanDavis)
	fmt.Println(ryanDavis.health)
}